src/stage1/ir.cpp

```diff
@@ -29850,23 +29850,28 @@ static IrInstGen *ir_analyze_ptr_cast(IrAnalyze *ira, IrInst* source_instr, IrIn
             return ira->codegen->invalid_inst_gen;
     }
 
-    if ((err = type_resolve(ira->codegen, dest_type, ResolveStatusZeroBitsKnown)))
+    if ((err = type_resolve(ira->codegen, dest_type, ResolveStatusSizeKnown)))
         return ira->codegen->invalid_inst_gen;
 
-    if ((err = type_resolve(ira->codegen, src_type, ResolveStatusZeroBitsKnown)))
+    if ((err = type_resolve(ira->codegen, src_type, ResolveStatusSizeKnown)))
         return ira->codegen->invalid_inst_gen;
 
-    if (safety_check_on &&
-        type_has_bits(ira->codegen, dest_type) &&
-        !type_has_bits(ira->codegen, if_slice_ptr_type))
-    {
+    size_t src_size = type_size_bits(ira->codegen, if_slice_ptr_type);
```

Comment:
AFAIR pointers are either `usize*8` or zero bits wide, is there some other edge case in between?

Reply:
The major case this intends to fix is an optional zero-sized pointer, which are 1 bit large. Do those not count as pointers? I know `?*allowzero T` doesn't count as a pointer type, but I wasn't sure if this case was considered the same way.

Reply:
Certainly in the past, pointers in C are not always a single machine word.   Think far and near pointers in the DOS era.  Also, the x32 mode for x86 CPUs also does weird things where pointers are 32-bit.   Java had/has a mode in the JVM where 32-bit pointers were used in 64-bit mode so registers could be 64-bit but pointers were smaller.  

The C specification goes to some lengths to make sure that the size of a pointer is not specified.   As long as `usize` is always the equivalent of C's `uintptr_t` then you can fit a pointer into it, but if Zig supports different memory models like the ones I noted above, pointers are not going to always be the same size.

Whether that is something that Zig wants to support different pointer sizes is a different story!

Reply:
> The major case this intends to fix is an optional zero-sized pointer, which are 1 bit large. Do those not count as pointers?

Those should probably decay to an optional type with no payload that in turn decay into a `u1` tag.

Reply:
@LemonBoy Yeah, but should they be allowed as the output/input type of @ptrCast?

Reply:
Casting "down" from a regular pointer to a zero-sized pointer should be probably allowed as that's safe while the other way around should be prohibited as that's a great way to shoot yourself in the foot.

Reply:
@LemonBoy #1469, an accepted proposal this is meant to close, disallows casting from a sized to zero-sized pointer. 

But should `*void` be able to be casted to `?*void`? They have different memory layouts.

What makes me think that it shouldn't is that you can't cast `*allowzero T` to a `?*allowzero T`, which sets a precedent.